### PR TITLE
ci: add linting & lint fixes

### PR DESCRIPTION
**Pull request by tac0turtle**
this pr brings linting to ci & lints the repo

there are two more errors, in places the repo uses src & c for `func (c *Chain)` to keep consistency either src or c should be chosen. I would prefer `c` are there are any preferences

### CODE REVIEW

**Comment by tac0turtle** [comment]
why is `-coverprofile coverage.out` for every test in the make file also the coverage isnt being used anywhere right?

**Comment by tac0turtle** [comment]
not sure why github actions arent running?

cc @jackzampolin

**Comment by jackzampolin** [comment]
@marbar3778 its cause you are pushing from a fork :( I've added you to the repo so you should be able to push your own branches now.

File: .github/workflows/gaia-tests.yml [hunk 1]
```diff
@@ -3,31 +3,29 @@ name: TESTING - gaia to gaia integration
...
-        key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
-        restore-keys: |
-          ${{ runner.os }}-go-
-    
-    # run tests
-    - name: run gaia tests
-      run: TEST_DEBUG=true make test-gaia
+      # run tests
+      - name: run gaia tests
+        run: make test-gaia
```
Comment (jackzampolin) [inline .github/workflows/gaia-tests.yml:31]:
```suggestion
        run: TEST_DEBUG=true make test-gaia
```

File: .github/workflows/mtd-tests.yml [hunk 2]
```diff
@@ -3,31 +3,22 @@ name: TESTING - mtd to ibc integration
...
-        key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
-        restore-keys: |
-          ${{ runner.os }}-go-
-    
-    # run tests
-    - name: run mtd tests
-      run: TEST_DEBUG=true make test-mtd
+      # run tests
+      - name: run mtd tests
+        run: make test-mtd
```
Comment (jackzampolin) [inline .github/workflows/mtd-tests.yml:24]:
```suggestion
        run: TEST_DEBUG=true make test-mtd
```

File: .github/workflows/gaia-tests.yml (hunk 1 above)
Comment (tac0turtle) [reply .github/workflows/gaia-tests.yml:31]: isn't this redundant as its set in the make file as well?

File: .github/workflows/rocketzone-tests.yml [hunk 3]
```diff
@@ -28,4 +28,4 @@ jobs:
 
       # run tests
       - name: run rocketzone tests
-        run: TEST_DEBUG=true make test-rocketzone
+        run: make test-rocketzone
```
Comment (jackzampolin) [inline .github/workflows/rocketzone-tests.yml:31]:
```suggestion
        run: TEST_DEBUG=true make test-rocketzone
```

File: cmd/config.go [hunk 4]
```diff
@@ -31,6 +31,8 @@ import (
 	"gopkg.in/yaml.v2"
 )
 
+const ORDERED = "ORDERED"
```
Comment (jackzampolin) [inline cmd/config.go]:
```suggestion
const (
  ORDERED = "ORDERED"
  UNORDERED = "UNORDERED"
)
```

File: cmd/config.go [hunk 5]
```diff
@@ -377,7 +380,7 @@ func overWriteConfig(cmd *cobra.Command, cfg *Config) error {
 			}
 
 			// overwrite the config file
-			err = ioutil.WriteFile(viper.ConfigFileUsed(), out, 0666)
+			err = ioutil.WriteFile(viper.ConfigFileUsed(), out, 0600)
```
Comment (jackzampolin) [inline cmd/config.go:396]: Why `600`?

Comment (tac0turtle) [reply cmd/config.go:396]: `G306: Poor file permissions used when writing to a new file`

File: .github/workflows/gaia-tests.yml (hunk 1 above)
Comment (jackzampolin) [reply .github/workflows/gaia-tests.yml:31]: Ah! Thats why! Thank you!

**Comment by tac0turtle** [comment]
> its cause you are pushing from a fork

how do you know if someone pushing "fixes" from a fork doesn't break everything?

**Comment by colin-axner** [comment]
what is the status of this pr? Can we get it merged in?

**Comment by tac0turtle** [comment]
yea this can be merged. If you are a member of this repo you may want to push it yourself to test ci. My invite expired so I never got around to testing it with ci

**Comment by colin-axner** [comment]
@marbar3778 do you know how we can update the build/tests to run on forks as well? Seems like a feature necessary for this repo. Is it because the repo is a fork itself?

**Comment by tac0turtle** [comment]
it seems to be running now. I'm not sure what changed.

File: test/test_setup.go [hunk 6]
```diff
@@ -179,18 +183,19 @@ func cleanUpTest(t *testing.T, testsDone <-chan struct{}, contDone chan<- struct
 	// remove all the docker containers
 	for i, r := range resources {
 		if err := pool.Purge(r); err != nil {
-			require.NoError(t, fmt.Errorf("Could not purge container %s: %w", r.Container.Name, err))
+			require.NoError(t, fmt.Errorf("cculd not purge container %s: %w", r.Container.Name, err))
```
Comment (colin-axner) [inline test/test_setup.go]:
```suggestion
			require.NoError(t, fmt.Errorf("could not purge container %s: %w", r.Container.Name, err))
```

File: relayer/packet-tx.go [hunk 7]
```diff
@@ -135,16 +136,16 @@ func (src *Chain) SendTransferBothSides(dst *Chain, amount sdk.Coin, dstAddr sdk
...
 	}
 
-	txs.Send(src, dst)
+	txs.Send(c, dst)
 	return nil
 }
 
 // SendTransferMsg initiates an ibs20 transfer from src to dst with the specified args
-func (src *Chain) SendTransferMsg(dst *Chain, amount sdk.Coin, dstAddr sdk.AccAddress, source bool) error {
+func (c *Chain) SendTransferMsg(dst *Chain, amount sdk.Coin, dstAddr fmt.Stringer, source bool) error {
```
Comment (colin-axner) [inline relayer/packet-tx.go:143]: why the change from sdk.AccAddress to fmt.Stringer?

**Review by colin-axner — COMMENTED** [review]
LGTM, left just a question on a change

**Comment by colin-axner** [comment]
> it seems to be running now. I'm not sure what changed.

I pushed your branch, so this is still an issue

File: relayer/packet-tx.go (hunk 7 above)
Comment (tac0turtle) [reply relayer/packet-tx.go:143]: this was due to the intefacer linter. I have seen false positives in the past with this liter but still find it useful for somethings. Would we want to disable it?

**Comment by tac0turtle** [comment]
I think its something in the settings, I'm not sure how to do it myself

File: relayer/packet-tx.go (hunk 7 above)
Comment (colin-axner) [reply relayer/packet-tx.go:143]: I think it should be fine for now

**Comment by colin-axner** [comment]
I'm not sure why the gaia-integration test is failing, I see `panic: test timed out after 10m0s` but i'm not sure what this is being caused by?

**Comment by colin-axner** [comment]
I think this is related how to the tests are being run here. I noticed locally I have to rerun `make test` otherwise running `make gaia-test` will result in an error from a dup function (I think it is being compiled twice or something?)

**Comment by tac0turtle** [comment]
> I think this is related how to the tests are being run here. I noticed locally I have to rerun `make test` otherwise running `make gaia-test` will result in an error from a dup function (I think it is being compiled twice or something?)

do the tests pass locally.

**Comment by colin-axner** [comment]
> do the tests pass locally.

no I get the same `panic: test timed out after 10m0s` issue. I think this is unrelated to what I said above, I'll go through the diffs again more closely sometime this week and see if there is something causing the tests to hang

File: relayer/relayMsgs.go [hunk 8]
```diff
@@ -45,8 +45,10 @@ func (r *RelayMsgs) IsMaxTx(msgLen, txSize uint64) bool {
 // Send sends the messages with appropriate output
 // TODO: Parallelize? Maybe?
 func (r *RelayMsgs) Send(src, dst *Chain) {
-	var msgLen, txSize uint64
-	var msgs []sdk.Msg
+	var (
+		msgLen, txSize uint64
+		msgs           = make([]sdk.Msg, len(r.Src))
```
Comment (colin-axner) [inline relayer/relayMsgs.go]: I think this should probably not pre alloc. msgs is used to batch send msgs that need to be relayed by using preset max tx length (in bytes) or send a max amount of msgs using msgLen, unfortunately msgLen is not always set. Also, it is used by r.Dst which may not have the same length as r.Src

File: relayer/query.go [hunk 9]
```diff
@@ -139,7 +139,7 @@ func (c *Chain) QueryClientConsensusState(srcHeight, srcClientConsHeight int64)
 		return clientTypes.NewConsensusStateResponse("notfound", nil, nil, 0), nil
 	}
 
-	var cs exported.ConsensusState
+	var cs clientExported.ConsensusState
```
Comment (colin-axner) [inline relayer/query.go]:
```suggestion
	var cs clientexported.ConsensusState
```
this is how we name it in the SDK

File: relayer/query.go [hunk 10]
```diff
@@ -217,7 +218,7 @@ func (c *Chain) QueryClientState() (*clientTypes.StateResponse, error) {
 		return nil, nil
 	}
 
-	var cs exported.ClientState
+	var cs clientExported.ClientState
```
Comment (colin-axner) [inline relayer/query.go]:
```suggestion
	var cs clientexported.ClientState
```

File: relayer/client-tx.go [hunk 11]
```diff
@@ -8,43 +8,43 @@ import (
...
-			src.logCreateClient(dst, dstH.GetHeight())
+		if c.debug {
+			c.logCreateClient(dst, dstH.GetHeight())
 		}
-		clients.Src = append(clients.Src, src.PathEnd.CreateClient(dstH, dst.GetTrustingPeriod(), src.MustGetAddress()))
+		clients.Src = append(clients.Src, c.PathEnd.CreateClient(dstH, dst.GetTrustingPeriod(), c.MustGetAddress()))
 	}
 
-	// Create client for src on dst if it doesn't exist
+	// Create client for c on dst if it doesn't exist
```
Comment (colin-axner) [inline relayer/client-tx.go]:
```suggestion
	// Create client for the source chain on destination chain if it doesn't exist
```

File: relayer/client-tx.go [hunk 12]
```diff
@@ -8,43 +8,43 @@ import (
 )
 
 // CreateClients creates clients for src on dst and dst on src given the configured paths
-func (src *Chain) CreateClients(dst *Chain) (err error) {
+func (c *Chain) CreateClients(dst *Chain) (err error) {
 	clients := &RelayMsgs{Src: []sdk.Msg{}, Dst: []sdk.Msg{}}
 
-	// Create client for dst on src if it doesn't exist
+	// Create client for dst on c if it doesn't exist
```
Comment (colin-axner) [inline relayer/client-tx.go]:
```suggestion
	// Create client for the destination chain on the source chain if it doesn't exist
```

File: .github/workflows/build.yml [hunk 13]
```diff
@@ -1,55 +1,55 @@
...
+  golangci:
+    name: golangci-lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: golangci/golangci-lint-action@master
+        with:
+          # Required: the version of golangci-lint is required and must be specified without patch version: we always use the latest patch version.
+          version: v1.27
+          args: --timeout 10m
```
Comment (colin-axner) [inline .github/workflows/build.yml:54]: is it possible this timeout is affecting the tests?

File: test/test_setup.go [hunk 14]
```diff
@@ -13,16 +13,16 @@ import (
...
 
 // spinUpTestChains is to be passed any number of test chains with given configuration options
 // to be created as individual docker containers at the beginning of a test. It is safe to run
 // in parallel tests as all created resources are independent of eachother
-func spinUpTestChains(t *testing.T, testChains ...testChain) Chains {
+func spinUpTestChains(t *testing.T, testChains ...testChain) ry.Chains {
 	var (
 		resources []*dockertest.Resource
-		chains    []*Chain
+		chains    = make([]*ry.Chain, len(testChains))
```
Comment (colin-axner) [inline test/test_setup.go:25]: this is the bug, using pre-alloc is not compatible with `append` usage

**Comment by colin-axner** [comment]
I'm running into this new issue locally:
```
  TestGaiaToGaiaStreamingRelayer: relayer_gaia_test.go:59: 
        	Error Trace:	relayer_gaia_test.go:59
        	Error:      	Received unexpected error:
        	            	RPC error -32603 - Internal error: 
        	            	parse error near Unknown (line 1 symbol 1 - line 1 symbol 1):
        	            	""
        	            	
        	            	TxSearch
        	            	github.com/tendermint/tendermint/rpc/client/http.(*baseRPCClient).TxSearch
        	            		/home/colinaxner/go/pkg/mod/github.com/tendermint/tendermint@v0.33.4/rpc/client/http/http.go:406
        	            	github.com/iqlusioninc/relayer/relayer.(*Chain).QueryTxs
        	            		/home/colinaxner/go/src/github.com/iqlusioninc/relayer/relayer/query.go:995
        	            	github.com/iqlusioninc/relayer/relayer.packetMsgFromTxQuery
        	            		/home/colinaxner/go/src/github.com/iqlusioninc/relayer/relayer/naive-strategy.go:266
        	            	github.com/iqlusioninc/relayer/relayer.(*NaiveStrategy).RelayPacketsOrderedChan
        	            		/home/colinaxner/go/src/github.com/iqlusioninc/relayer/relayer/naive-strategy.go:207
        	            	github.com/iqlusioninc/relayer/relayer.RunStrategy
        	            		/home/colinaxner/go/src/github.com/iqlusioninc/relayer/relayer/strategies.go:77
        	            	github.com/iqlusioninc/relayer/test.TestGaiaToGaiaStreamingRelayer
        	            		/home/colinaxner/go/src/github.com/iqlusioninc/relayer/test/relayer_gaia_test.go:58
        	            	testing.tRunner
        	            		/usr/local/go/src/testing/testing.go:992
        	            	runtime.goexit
        	            		/usr/local/go/src/runtime/asm_amd64.s:1373
        	Test:       	TestGaiaToGaiaStreamingRelayer
```

**Comment by colin-axner** [comment]
I removed the `-race` condition test because tests won't pass until the race condition is addressed and that is outside the scope of this pr, an issue has been opened #276
